### ISSUE TITLE
String.indexOf() cannot match phrases with variable whitespace

### ISSUE THREAD

**Issue by rpedela** [mozilla/pdf.js#7355]
`String.indexOf()` is being used for phrase search. There are two scenarios where this will fail because of whitespace:
1. Justified text, very common in legal documents, which can result in variable amount of whitespace between words.
2. Currently spaces are not added between lines therefore phrases that span multiple lines will not be found. See #5783 for more information.

Even if 2 is fixed, we still have to deal with 1 for phrase search. I have written a JS function that is equivalent to `String.indexOf()` but ignores `" "` while matching. This function also returns the end index since the length of `searchValue` cannot be trusted because the amount of whitespace between `text` and `searchValue` may differ. It could be modified to handle all whitespace characters, but I think it is unnecessary in this case since the text layer converts all whitespace to `" "`.

``` js
function indexOfIgnoreSpace(text, searchValue, fromIndex) {

  var begin = 0;
  var end = -1;
  var fragment = searchValue.slice(0).split(' ').join(''); // remove whitespace

  if (!fromIndex) {
    fromIndex = 0;
  }

  for (var i = fromIndex; i < text.length; i++) {

    // do not start matching on whitespace
    if (text[i] === ' ') {
      continue;
    }

    var index = 0;
    begin = i;

    for (var j = i; j < text.length; j++) {

      if (text[j] === ' ') {
          continue;
      }

      if (text[j] !== fragment[index]) {
          break;
      }

      index++;
      if (index === fragment.length) {
          end = j + 1;
          break;
      }
    }

    if (end !== -1) {
      return {
        begin: begin,
        end: end
      };
    }
  }

  return null;
}
```

**Comment by luomancs** [comment]
hi,
where should I put your function in the code so that it can resolve the space mismatch problem? thank you.

**Comment by calixteman** [comment]
It's partially fixed thanks to af4dc55 which helps to detect EOL.
And this PR: #13261 should definitely fix the problem.
